netket/experimental/operator/_particle_number_conserving_fermionic/_operators.py

```diff
@@ -0,0 +1,341 @@
+from netket.operator import FermionOperator2ndJax
+from netket.utils.optional_deps import import_optional_dependency
+
+from .._pyscf_utils import (
+    TV_from_pyscf_molecule,
+    to_desc_order_sparse,
+    compute_pyscf_integrals,
+)
+
+from .._normal_order_utils import to_normal_order, to_normal_order_sector
+from ._conversion import (
+    fermiop_terms_to_sites_daggers_weights,
+    fermiop_terms_to_sites_sectors_daggers_weights,
+    to_fermiop_helper,
+)
+from ._operator_data import (
+    collect_ops,
+    sparse_arrays_to_coords_data_dict,
+    prepare_operator_data_from_coords_data_dict,
+    prepare_operator_data_from_coords_data_dict_spin,
+    to_coords_data_sector,
+)
+from ._kernels import get_conn_padded_pnc, get_conn_padded_pnc_spin
+
+
+@struct.dataclass
+class ParticleNumberConservingFermioperator2ndJax(DiscreteJaxOperator):
+    """
+    Particle-number conserving fermionc operator
+    H = w + Σ_ij w_ij c_i^† c_j + Σ_ijkl w_ijkl  c_i^† c_j^† c_k c_l + Σ_ijklmn w_ijklmn c_i^† c_j^† c_k^† c_l c_m c_n + ...
+
+    Version without spin.
+
+    To be used with netket.hilbert.SpinOrbitalFermions with a fixed number of fermions.
+
+    It uses a custom sparse internal representation,
+    please refer to the docstrings of prepare_data and prepare_data_diagonal for details.
+
+    We provide several factory methods to create this operator:
+        - ParticleNumberConservingFermioperator2ndJax.from_fermiop:
+               Conversion form FermionOperator2nd/FermionOperator2ndJax
+        - ParticleNumberConservingFermioperator2ndJax.from_sparse_arrays:
+                From sparse arrays (w, w_ij, w_ijkl, w_ijklmn, ...)
+        - ParticleNumberConservingFermioperator2ndJax.from_pyscf_molecule:
+                From pyscf
+
+    Furthermore it can be converted to FermionOperator2nd/FermionOperator2ndJax using the .to_fermiop method.
+    """
+
+    # factory methods for internal use only:
+    # - ParticleNumberConservingFermioperator2ndJax._from_coords_data_normal_order:
+    #         From tuples of (sites, daggers, weights) representing w, w_ij, ...
+    #         where only the lower triangular part is nonzero
+    # - ParticleNumberConservingFermioperator2ndJax._from_sparse_arrays_normal_order:
+    #         From sparse arrays (w, w_ij, w_ijkl, w_ijklmn) where i>=j, i>=j>=k>=l etc,
+    #         and only the lower triangular part is nonzero
+
+    _hilbert: SpinOrbitalFermions = struct.field(pytree_node=False)
+    _operator_data: PyTree  # custom sparse internal representation
+
+    def get_conn_padded(self, x):
+        return get_conn_padded_pnc(self._operator_data, x, self._hilbert.n_fermions)
+
+    @property
+    @jax.jit
+    def max_conn_size(self):
+        x = jax.ShapeDtypeStruct((1, self._hilbert.size), dtype=jnp.uint8)
+        _, mels = jax.eval_shape(self.get_conn_padded, x)
+        return mels.shape[-1]
+
+    @property
+    def dtype(self):
+        return NotImplemented
```

Comment:
This should be the dtype of the matrix elements, so it's the dtype of the values within operator data.